src/main/java/org/craftercms/engine/util/servlet/ConfigAwareProxyServlet.java

```diff
@@ -84,4 +110,105 @@ protected void copyResponseHeaders(HttpResponse proxyResponse, HttpServletReques
+        URL url = new URL(proxyRequestUri);
+
+        Socket sock = new Socket(url.getHost(), url.getPort());
+        boolean closeSocket = false;
+        try {
+            var sockIn = sock.getInputStream();
+            var sockOut = sock.getOutputStream();
+
+            StringBuilder req = new StringBuilder(512);
+            req.append("GET " + url.getFile()).append(" HTTP/1.1");
+            logger.debug("> Websocket|", req);
+            req.append("\r\n");
+            var headerNames = servletRequest.getHeaderNames();
+            while (headerNames.hasMoreElements()) {
+                var name = headerNames.nextElement();
+                String header = servletRequest.getHeader(name);
+                logger.debug("> Websocket| {}:{}", name, header);
+                req.append(name + ": " + header + "\r\n");
+            }
+            req.append("\r\n");
+
+            sockOut.write(req.toString().getBytes(StandardCharsets.UTF_8));
+            sockOut.flush();
+
+            StringBuilder responseBytes = new StringBuilder(512);
+            int b = 0;
+            while (b != -1) {
+                b = sockIn.read();
+                if (b != -1) {
+                    responseBytes.append((char) b);
+                    var len = responseBytes.length();
+                    if (len >= 4
+                            && responseBytes.charAt(len - 4) == '\r'
+                            && responseBytes.charAt(len - 3) == '\n'
+                            && responseBytes.charAt(len - 2) == '\r'
+                            && responseBytes.charAt(len - 1) == '\n'
+                    ) {
+                        break;
+                    }
+                }
+            }
+
+            String[] rows = responseBytes.toString().split("\r\n");
+
+            String response = rows[0];
+            logger.debug("< Websocket| '{}'", response);
+
+            int idx1 = response.indexOf(' ');
+            int idx2 = response.indexOf(' ', idx1 + 1);
+
+            for (int i = 1; i < rows.length; i++) {
+                String line = rows[i];
+                int idx3 = line.indexOf(":");
+                String k = line.substring(0, idx3);
+                String headerField = line.substring(idx3 + 2);
+                logger.debug("< Websocket| {}:{}", k, headerField);
+                servletResponse.setHeader(k, headerField);
+            }
+
+            int respCode = Integer.parseInt(response.substring(idx1 + 1, idx2));
+            if (respCode != HttpServletResponse.SC_SWITCHING_PROTOCOLS) {
+                servletResponse.setStatus(respCode);
+                servletResponse.flushBuffer();
+                logger.debug("< Websocket| Flush");
+                closeSocket = true;
+            } else {
+                var upgradeHandler = servletRequest.upgrade(WsUpgradeHandler.class);
+                upgradeHandler.preInit(exec, sockIn, sockOut, sock);
```

Comment:
This just calls preInit, what about the init method? The variable is not used anywhere else.

Reply:
@avasquez614 `WsUpgradeHandler` implement `HttpUpgradeHandler` and the `init` method will be called when the handshake has been completed. The `preInit` method is just to pass the input stream and output stream of the socket connection between Engine <==> Websocket server so that we can forward the data among Web Client <==> Engine as the Proxy <==> Websocket Server

In our case, we have completed the handshake and checked the status code is `101` in https://github.com/craftercms/engine/pull/682/files#diff-44b6059479d4bd49c012bbcd46e18e48b5c2281d8671873ee6cf2ce6e524021aR160 so the `init` will be called.
Post here the documentation for `HttpUpgradeHandler`: https://tomcat.apache.org/tomcat-9.0-doc/servletapi/javax/servlet/http/HttpUpgradeHandler.html#init(javax.servlet.http.WebConnection)
> **
     * It is called once the HTTP Upgrade process has been completed and
     * the upgraded connection is ready to start using the new protocol.
     *
     * @param wc the WebConnection object associated to this upgrade request
     */